Tools/clinic/clinic.py

```diff
@@ -850,6 +873,57 @@ def render(
                 function = o
         return self.render_function(clinic, function)
 
+    def deprecate_positional_use(
+            self,
+            func: Function,
+            params: dict[int, Parameter],
+    ) -> str:
+        assert len(params) > 0
+        names = [repr(p.name) for p in params.values()]
+        first_pos, first_param = next(iter(params.items()))
+        last_pos, last_param = next(reversed(params.items()))
+
+        # Pretty-print list of names.
+        pstr = pprint_words(names)
+
+        # For now, assume there's only one deprecation level.
+        assert first_param.deprecated_positional == last_param.deprecated_positional
+        thenceforth = first_param.deprecated_positional
+        assert thenceforth is not None
+
+        # Format the preprocessor warning and error messages.
+        assert isinstance(self.cpp.filename, str)
+        source = os.path.basename(self.cpp.filename)
+        major, minor = thenceforth
+        cpp_message = (
+            f"In {source}, update parameter(s) {pstr} in the clinic "
+            f"input of {func.full_name!r} to be keyword-only."
+        )
+        # Format the deprecation message.
+        if len(params) == 1:
+            condition = f"nargs == {first_pos+1}"
+            depr_message = (
+                f"Passing {first_pos+1} positional arguments to "
+                f"{func.full_name}() is deprecated. Parameter {pstr} will "
+                f"become a keyword-only parameter in Python {major}.{minor}."
+            )
+        else:
+            condition = f"nargs > {first_pos} && nargs <= {last_pos+1}"
+            depr_message = (
+                f"Passing more than {first_pos} positional arguments to "
+                f"{func.full_name}() is deprecated. Parameters {pstr} will "
+                f"become keyword-only parameters in Python {major}.{minor}."
+            )
```

Comment:
Some edge cases where this wouldn't generate ideal deprecation messages:

## Problematic case 1

`def f(x): ...` -> `def f(*, x): ...`. If I understand correctly, the deprecation message here would be

> Passing 1 positional arguments to f() is deprecated. Parameter x will become a keyword-only parameter in Python 3.14

I think we should special-case functions with only one parameter, so that the error message would be:

> Passing positional arguments to f() is deprecated. Parameter x will become a keyword-only parameter in Python 3.14

## Problematic case 2

`def g(x, y): ...` -> `def g(*, x, y): ...`. If I understand correctly, the error message here would be:

> Passing more than 0 positional arguments to g() is deprecated. Parameters x and y will become keyword-only parameters in Python 3.14.

A better deprecation message would be:

> Passing positional arguments to g() is deprecated. Parameters x and y will become keyword-only parameters in Python 3.14.

## Problematic case 3:

`def h(x, y, z): ...` -> `def h(x, *, y, z): ...`. If I understand correctly, the error message here would be:

> Passing more than 1 positional arguments to h() is deprecated. Parameters y and z will become keyword-only parameters in Python 3.14.

A better deprecation message would be:

> Passing more than 1 positional argument to h() is deprecated. Parameters y and z will become keyword-only parameters in Python 3.14.

(Singular "argument", rather than plural "arguments")

Reply:
Oh, sorry, just saw your message in https://github.com/python/cpython/pull/95151#issuecomment-1666974001! Yes, we can leave these edge cases for now :)

Reply:
Yeah, I suggest we create a pretty-print-deprecation-warning helper in a follow-up PR. There's enough churn on this PR already :)

Reply:
I think I got most of them in bde0f90d4b9cbd3ec590fdfaf5d7192b54697db7